collaboration/sig-core/proposals/kyma-controller-manager.md

```diff
@@ -0,0 +1,181 @@
+# Consolidate Kyma controllers inside a single controller manager
+
+Created on 2019-10-11 by Antoine Cotten (@antoineco).
+
+## Status
+
+DRAFT. Proposed on 2019-10-11.
+
+## Motivation
+
+Kyma is composed of multiple components that share a lot of architectural patterns, yet are written in isolation
+following different engineering standards and using heterogeneous SDKs and libraries, before eventually being released
+as _one product_.
+
+Most of these components involve one or more Kubernetes controllers that synchronize API objects using identical
+building blocks: watches, work queues, reconciliation loops, etc. Every controller instance maintains its own local
+cache of several Kubernetes resources, which can result in a significant memory footprint in productive environments.
+Besides, each of these controllers is operated in a similar fashion by nature: as a single instance or in an
+active-passive setup.
+
+These similarities represent an opportunity for us to consolidate Kyma controllers and run them under the umbrella of a
+single _manager_, so that they can reuse common libraries, maintain a consistent level of compatibility with the
+underlying Kubernetes layer, share memory efficiently, and avoid the fragmentation we are currently observing in the
+project.
+
+This mode of operation is directly inspired by the Kubernetes [controller-manager][k8s-cm].
+
+## Goals
+
+This initiative aims at improving three aspects:
+
+**Maintainability**
+
+Streamline the codebase regarding controllers to harmonize the different implementation aspects. (_"Common opinion"_)
+
+**Footprint**
+
+Reduce the resource consumption by sharing internal API object caches whenever possible. (_"Store once"_)
+
+**Operability**
+
+Improve operational aspects by reducing the number of controllers that have to be deployed separately. (_"Run once"_)
+
+## Constraints
+
+It must be possible to opt out of each individual controller inside the manager to keep Kyma modular. This could be
+achieved using startup flags and/or a ConfigMap.
+
+A controller ("operator") SDK should be agreed upon by all code owners. [kyma#4327][kyma4327] initiated
+a discussion around that topic, but no conclusion has been reached at the time of writing. For the initial version of
+this document, we assume that [`kubebuilder`][kb] will be our SDK of choice. A few reasons to believe this assumption is
+safe to make:
```

Comment:
here is my comment about going with the kubebuilder:
https://github.com/kyma-project/kyma/issues/4327#issuecomment-541596179

Some time ago I advocated the kubebuilder approach. Also, some of Gopher components are written with kubebuilder. But after some time I've changed my opinion about that and I've described that in the linked issue.

Reply:
I agree with your analysis. I'm don't agree with all of kubebuilder's opinions myself, we have to strike a good balance and be ready to fill some gaps on our own. I just didn't want to send a proposal that suggests to agree on a SDK but doesn't take any position, because the discussion would be harder to start without an initial assumption.

Reply:
A real-life scenario where kubebuilder has bitten us: one of CBS dependencies was forcing kubernetes 1.15, but kubebuilder was using older controller-runtime at that time (pre 2.0 times). We ended up with a feature postponed until kubebuilder 2.0

Reply:
At that point it's clear kubebuilder is a bad choice for us, so this implementing this proposal will require a decision for kyma-project/kyma#4327.

Reply:
Removed references to kubebuilder.